src/Transform/ONNX/ConstProp.cpp

```diff
@@ -243,229 +194,125 @@ ONNXConstantOp createConstantOpAndStoreBufferPtr(
+auto combinerOfElementwiseBinaryOp(BType operandsBType) {
+  using Combiner = std::function<WideNum(WideNum, WideNum)>;
+  return dispatchByBType(operandsBType, [](auto btype) -> Combiner {
+    using W = WideBType<btype>;
+    using OpImpl =
+        ElementWiseBinaryOpImpl<ElementwiseBinaryOp, typename W::type>;
+    return [](WideNum lhs, WideNum rhs) -> WideNum {
+      return W::pack(OpImpl::impl(W::unpack(lhs), W::unpack(rhs)));
+    };
+  });
 }
 
 /// Do element-wise binary calculation of 'lhs' and 'rhs' values and create an
 /// ONNXConstantOp for the result.
 template <typename ElementwiseBinaryOp>
-Value ConstPropElementwiseBinary(
-    PatternRewriter &rewriter, Value replacingValue, Value lhs, Value rhs) {
-  Type elementType =
-      replacingValue.getType().cast<ShapedType>().getElementType();
-  ArrayRef<int64_t> lhsShape = lhs.getType().cast<ShapedType>().getShape();
-  ArrayRef<int64_t> rhsShape = rhs.getType().cast<ShapedType>().getShape();
-  ArrayRef<int64_t> outputShape =
-      replacingValue.getType().cast<ShapedType>().getShape();
-
-  // Get lhs and rhs values.
-  char *lhsArray = getArrayFromAttributeOrBuffer(rewriter, lhs.getDefiningOp());
-  char *rhsArray = getArrayFromAttributeOrBuffer(rewriter, rhs.getDefiningOp());
-
-  // Do calculation.
-  // Use maximum size (double or int64_t) to avoid the precision loss.
-  char *resArray =
-      allocateBufferFor(replacingValue.getType(), /*useMaxSize=*/true);
-  if (elementType.isa<FloatType>()) {
-    // Use double to avoid the precision loss during computation.
-    IterateConstPropElementwiseBinary<ElementwiseBinaryOp, double>(
-        lhsArray, rhsArray, lhsShape, rhsShape, resArray, outputShape);
-  } else if (elementType.isa<IntegerType>()) {
-    // Use int64_t to avoid the precision loss during computation.
-    IterateConstPropElementwiseBinary<ElementwiseBinaryOp, int64_t>(
-        lhsArray, rhsArray, lhsShape, rhsShape, resArray, outputShape);
-  } else
-    llvm_unreachable("Unknown data type");
-
-  // Construct a new ONNXConstantOp.
-  ONNXConstantOp res =
-      createConstantOpAndStoreBufferPtr(rewriter, replacingValue, resArray);
-
-  return res.getResult();
+Value ConstPropElementwiseBinary(PatternRewriter &rewriter,
+    Value replacingValue, Value lhsValue, Value rhsValue) {
+  ConstPropCounters::count("ElementwiseBinary", {lhsValue, rhsValue});
+  Type replacingType = replacingValue.getType().cast<ShapedType>();
+
+  ElementsAttrBuilder elementsBuilder(rewriter.getContext());
+  DisposableElementsAttr lhs =
+      getConstValueAsDisposableElements(elementsBuilder, lhsValue);
+  DisposableElementsAttr rhs =
+      getConstValueAsDisposableElements(elementsBuilder, rhsValue);
+  BType operandsBType = lhs.getBType();
+  assert(operandsBType == rhs.getBType());
+  ElementsAttr resultElements = elementsBuilder.combine(lhs, rhs, replacingType,
+      combinerOfElementwiseBinaryOp<ElementwiseBinaryOp>(operandsBType));
```

Comment:
I would move `combinerOfElementwiseBinaryOp` into `ElementsAttrBuilder` and provide, say `ElementsAttrBuilder.binaryOp<ElementwiseBinaryOp>`, which  is the same as you did for `reshape`, `transpose`. By that way, we don't need to expose `BType` here.

Reply:
all your 4 suggestions are good, I'll go to work on all of them and let you know if I run into any obstacles